packages/template-retail-react-app/app/contexts/index.js

```
@@ -27,23 +29,103 @@ import PropTypes from 'prop-types'
+            return storageString
+        }
+        let res
+        try {
+            res = await api.shopperProducts.getCategory({
+                parameters: {
+                    id,
+                    levels
+                }
+            })
+        } catch (e) {
+            console.error(e)
+        }
+        return res
+    }
+
+    useEffect(() => {
+        // Server side, we only fetch level 0 categories, for performance, here
+        // we request the remaining two levels of category depth
+        Promise.all(
+            root?.[itemsKey]?.map(async (cat) => {
+                // check localstorage first for this data to help remediate O(n) server
+                // load burden where n = top level categories
+                const storedCategoryData = JSON.parse(
+                    window?.localStorage?.getItem(`${LOCAL_STORAGE_PREFIX}${cat?.id}-${locale}`)
+                )
+                if (storedCategoryData) {
+                    return storedCategoryData
+                } else {
+                    try {
+                        const res = await fetchCategoryNode(cat?.id, 2)
+                        // store fetched data in local storage for faster access / reduced server load
+                        window?.localStorage?.setItem(
+                            `${LOCAL_STORAGE_PREFIX}${cat?.id}-${locale}`,
+                            JSON.stringify({
+                                ...res,
+                                fetchTime: Date.now()
+                            })
+                        )
+                        return res
+                    } catch (e) {
+                        return e
+                    }
+                }
+            })
+        ).then((data) => {
+            const newTree = {
+                ...root,
+                [itemsKey]: data
+            }
+            setRoot(newTree)
+        })
+        .catch((err) => console.log(err))
```

This will log the error to the browser console without any context, and then do nothing. Is that what we want? Do we have a fallback / error state that the menus can show?

---

https://github.com/SalesforceCommerceCloud/pwa-kit/pull/875#discussion_r1073963614 - Regarding this issue and the one referenced, would it be better to `throw err` ?

---

They're different scenarios. For the other comment, I don't think we want error handling at that layer (don't catch/rethrow), just let it bubble up to whatever error handling we do have.

I'm not sure what the proper solution is, here. Depends on what happens when the category data requests fail. Probably not log and ignore, which is the current state. Maybe just let the error bubble up to react and then have proper error handling in a follow up ticket.